kube/src/api/metadata.rs

```diff
@@ -75,14 +75,44 @@ pub trait Resource {
 
     /// Metadata that all persisted resources must have
     fn meta(&self) -> &ObjectMeta;
+    /// Metadata that all persisted resources must have
+    fn meta_mut(&mut self) -> &mut ObjectMeta;
+}
+
+/// Helper methods for resources.
+pub trait ResourceExt: Resource {
     /// The name of the resource
-    fn name(&self) -> String;
+    fn name(&self) -> Option<String>;
+    /// Returns the name of the resource, panicking if it is
+    /// missing. Use this function if you know that name is set, for example
+    /// when resource was received from the apiserver.
+    /// Because of `.metadata.generateName` field, in other contexts name
+    /// may be missing.
+    fn expect_name(&self) -> String;
```

Comment:
Maybe `name_unchecked`?

Reply:
I was going to advocate for hiding it like we did before, but after testing it against quick `generateName` variants, and seeing that it's a common thing (and not just an api wart that could have been an admission controller), we kind of have to deal with it.

- The response back when you create with `generateName` does not have `metadata.name`.
- The first ADDED event from watch with `generateName` does not have `metadata.name` (so a plain `watcher` actually crashes, opening a bug)